arch/risc-v/src/mpfs/mpfs_corepwm.c

```diff
@@ -0,0 +1,787 @@
+ *
+ * Input Parameters:
+ *   priv - A reference to the lower half PWM driver state structure
+ *   info - A reference to the characteristics of the pulsed output
+ *
+ * Returned Value:
+ *   Zero on success; a negated errno value on failure
+ *
+ ****************************************************************************/
+
+static int pwm_timer(FAR struct mpfs_pwmtimer_s *priv,
+                     FAR const struct pwm_info_s *info)
+{
+  int      i;
+
+  /* Calculated values */
+
+  /* TODO: We might need to calculate prescaler on some rare cases,
+   * for now hardcoded to 0
+   */
+
+  uint32_t prescaler = 0;
+
+  uint32_t period;
+
+  DEBUGASSERT(priv != NULL && info != NULL);
+  DEBUGASSERT(info->frequency > 0);
+
+  /* CorePWM FPGA block can be configured to be with either 8, 16, or 32 bit
+   * registers width. Minimally PWM functionality is set up by two registers:
+   * PRESCALE and PERIOD which are common to all channels. Up to 16 channels
+   * may be configured in use. Clock used by the block may be selected in
+   * design phase an on Icicle Kit reference design version 21.04 has at
+   * least the following clock signals to choose from the Clocks_and_Resets
+   * block: 125MHz, 100MHz, 75MHz, 62.5MHz, 50MHz, and 25MHz.
+   *
+   * For now only 32 the bit configuration is supported.
+   * TODO: Add 8 and 16 bit width support
+   *
+   * There are many combinations of prescaler and period registers, but the
+   * best will be the one that has the smallest prescaler value. That is the
+   * solution that should give us the most accuracy in the pwm control.
+   *
+   * Example for clk = 25MHz, prescale 0 and 32 bit wide registers:
+   *   PWM period granularity PWM_PG = (PRESCALE + 1) / pwmclk =
+   *   40 ns × 1 = 40 ns, so the smallest step is 40ns
+   *   pwmclk = clk / (PRESCALE + 1) = 25,000,000 / (PRESCALE + 1) =
+   *     25,000,000
+   *
+   *    For desired output frequency of 50Hz and using PRESCALE of 0:
+   *    PERIOD = pwmclk / frequency = 25,000,000 / 50 = 500,000
+   */
+
+  pwminfo("PWM%u frequency: %u PWMCLK: %u prescaler: %u\n",
+          priv->pwmid, info->frequency, priv->pwmclk, prescaler);
+
+  /* Set the reload and prescaler values */
+
+  period = priv->pwmclk / info->frequency;
+
+  pwm_putreg(priv, MPFS_COREPWM_PERIOD_OFFSET, period);
+  pwm_putreg(priv, MPFS_COREPWM_PRESCALE_OFFSET, prescaler);
+
+  /* Handle channel specific setup */
+
+  for (i = 0; i < CONFIG_PWM_NCHANNELS; i++)
+    {
+      ub32_t    duty;
+      uint8_t   channel;
+      uint32_t  neg_edge;
+
+      channel   = info->channels[i].channel;
+
+      /* Duty defined as fraction of 65536, i.e. a value of 1 to 65535
+       * corresponding to a duty cycle of 0.000015 - 0.999984
+       */
+
+      duty      = ub16toub32(info->channels[i].duty);
+      neg_edge  = b32toi(duty * period + b32HALF);
+
+      if (channel == 0)   /* A value of zero means to skip this channel */
```

Comment:
Note that PR #4227 added an option to break these loops when `channel == -1`, it looks like it was missed here.

Reply:
Thanks for pointing this out!

Reply:
PR https://github.com/apache/incubator-nuttx/pull/5091 to fix.